workflow_scheduler.go

```diff
@@ -26,32 +42,95 @@ type ScheduleDescription struct {
 	WinEventSubscribeQueries [][]string `json:"win_event_subscribe_queries"`
 }
 
-/** Class that represents a query for subscribing to a windows event. */
-type WinEventSubscribeQuery struct {
-	channel string
-	query string
+/** Class that holds information used in the subscription callback function. */
+type CallbackContext struct {
+	workflow workflows.InterfaceWorkflow
+	workflowContext *nodes.WorkflowContext
+}
+
+func (workflowScheduler *WorkflowScheduler) SubscriptionCallback(Action wevtapi.EVT_SUBSCRIBE_NOTIFY_ACTION, UserContext win32.PVOID, Event wevtapi.EVT_HANDLE) uintptr {
+	var CStringGuid *C.char = (*C.char)(unsafe.Pointer(UserContext))
+	var guid string = C.GoStringN(CStringGuid, 32)
+	var callbackContext *CallbackContext = workflowScheduler.guidToContext[guid]
+
+	switch Action {
+		case wevtapi.EvtSubscribeActionError:
+			workflowScheduler.logger.Println("Win event subscription failed.")
+			return 0
+		case wevtapi.EvtSubscribeActionDeliver:
+			UTF16EventXML, err := wevtapi.EvtRenderXML(Event)
+			if err != nil {
+				workflowScheduler.logger.Println("Error converting event to XML:", err)
+			}
+			var eventXML string = win32.UTF16BytesToString(UTF16EventXML)
+
+			var event utils.EventFromXML = utils.NewUtils().ParseEventXML(eventXML)
+			var nowTime time.Time = time.Now()
+
+			// We use the start of today because the time defaults to 00:00 in event.TimeCreated.
+			var startOfToday time.Time = time.Date(nowTime.Year(), nowTime.Month(), nowTime.Day(), 0, 0, 0, 0, event.TimeCreated.Location())
+
+			var age float64 = startOfToday.Sub(event.TimeCreated).Hours() / float64(24)
+			if (age > configs.MAX_AGE_TO_PROCESS_WIN_EVTS_IN_DAYS) {
+				workflowScheduler.logger.Println("Event flagged as too old. Age:", age)
+				return uintptr(0)
+			}
+
+			callbackContext.workflowContext = nodes.NewWorkflowContext()
+			callbackContext.workflowContext.Seed = eventXML
+			callbackContext.workflowContext.Provider = event.Provider
+			callbackContext.workflowContext.EventID = event.EventID
+			callbackContext.workflowContext.EventRecordID = event.EventRecordID
+
+			// Create a goroutine to run the workflow included in the callback context.
+			go callbackContext.workflow.Run(callbackContext.workflow, callbackContext.workflowContext)
+		default:
+			workflowScheduler.logger.Println("encountered error during callback: unsupported action code %x", uint16(Action))
+	}
+	return uintptr(0)
+}
+
+func (workflowScheduler *WorkflowScheduler) storeCallbackContext(context *CallbackContext) string {
+	var uuidWithHyphen uuid.UUID = uuid.New()
+    var uuid string = strings.Replace(uuidWithHyphen.String(), "-", "", -1)
```

Comment:
Why remove the "-"? we're passing it as a string to the callback anyways, right?

Reply:
I saw this being done in the Python guid example where they use .hex() to remove the dashes. Thats true, I have removed it now.